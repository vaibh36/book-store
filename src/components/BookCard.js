import React, { useState } from "react";
import {
  Card,
  CardContent,
  Typography,
  IconButton,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Button,
  Stack,
} from "@mui/material";
import InfoIcon from "@mui/icons-material/Info";
import { useDispatch } from "react-redux";
import { toggleReadStatus } from "../store";
import { Edit } from "@mui/icons-material";
import { useNavigate } from "react-router-dom";

const BookCard = ({ book }) => {
  const [dialogOpen, setDialogOpen] = useState(false);
  const [selectedBook, setSelectedBook] = useState(null);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleDialogOpen = () => {
    setDialogOpen(true);
  };

  const handleDialogClose = () => {
    setSelectedBook(null);
    setDialogOpen(false);
  };

  return (
    <>
      <Card
        variant="outlined"
        sx={{
          height: "200px",
          minWidth: "250px",
        }}
      >
        <CardContent>
          <Typography
            variant="h6"
            component="div"
            noWrap
            sx={{
              overflow: "hidden",
              textOverflow: "ellipsis",
              whiteSpace: "nowrap",
            }}
          >
            {book.title}
          </Typography>
          <Typography variant="body2" color="text.secondary">
            by {book.author}
          </Typography>
          <Typography variant="body2" color="text.secondary">
            {book.read ? "Read" : "Unread"}
          </Typography>
          <Stack flexDirection={"row"} alignItems={"center"}>
            <Button
              data-testid="read__btn"
              variant="contained"
              color={book.read ? "primary" : "default"}
              onClick={() => {
                dispatch(toggleReadStatus(book.id));
              }}
              sx={{
                textTransform: "none",
                marginTop: 1,
              }}
            >
              {book.read ? "Mark as Unread" : "Mark as Read"}
            </Button>
            <IconButton
              data-testid="info__icon"
              color="primary"
              onClick={() => {
                setSelectedBook(book);
                handleDialogOpen();
              }}
              sx={{ float: "right" }}
            >
              <InfoIcon />
            </IconButton>
            <IconButton
              color="primary"
              onClick={() => {
                navigate(`/books/${book?.id}`, {
                  state: {
                    id: book?.id,
                    author: book?.author,
                    price: book?.price,
                    read: book?.read,
                    title: book?.title,
                  },
                });
              }}
              sx={{ float: "right" }}
            >
              <Edit />
            </IconButton>
          </Stack>
        </CardContent>
      </Card>

      <Dialog open={dialogOpen} onClose={handleDialogClose}>
        <DialogTitle>Book Information</DialogTitle>
        <DialogContent>
          <Typography variant="body1" gutterBottom>
            <strong>Title:</strong> {selectedBook?.title}
          </Typography>
          <Typography variant="body1" gutterBottom>
            <strong>Author:</strong> {selectedBook?.author}
          </Typography>
          <Typography variant="body1" gutterBottom>
            <strong>Price:</strong> {selectedBook?.price}
          </Typography>
          <Typography variant="body1">
            <strong>Status:</strong> {selectedBook?.read ? "Read" : "Unread"}
          </Typography>
        </DialogContent>
        <DialogActions>
          <Button
            data-testid="close__btn"
            onClick={handleDialogClose}
            color="primary"
            variant="contained"
          >
            Close
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default BookCard;
