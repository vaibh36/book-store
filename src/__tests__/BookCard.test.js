import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import BookCard from "../components/BookCard";

import configureStore from "redux-mock-store";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  BrowserRouter: ({ children }) => <div>{children}</div>,
  Route: ({ children }) => <div>{children}</div>,
}));
const initialState = {};

const mockStore = configureStore([]);

describe("BookCard Component", () => {
  const mockBook = {
    id: 1,
    title: "Test Book",
    author: "Test Author",
    price: "$10.00",
    read: false,
  };

  test("snapshot test case", () => {
    const wrapper = render(
      <Provider store={mockStore(initialState)}>
        <MemoryRouter>
          <BookCard book={mockBook} />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
  test("open dialog test", async () => {
    render(
      <Provider store={mockStore(initialState)}>
        <MemoryRouter>
          <BookCard book={mockBook} />
        </MemoryRouter>
      </Provider>
    );
    const infoIcon = await screen.findByTestId("info__icon");
    fireEvent.click(infoIcon);
  });
  test("close btn click", async () => {
    render(
      <Provider store={mockStore(initialState)}>
        <MemoryRouter>
          <BookCard book={mockBook} />
        </MemoryRouter>
      </Provider>
    );
    const infoIcon = await screen.findByTestId("info__icon");
    fireEvent.click(infoIcon);
    const closeBtn = await screen.findByTestId("close__btn");
    fireEvent.click(closeBtn);
  });
  test("read unread toggle click", async () => {
    render(
      <Provider store={mockStore(initialState)}>
        <MemoryRouter>
          <BookCard book={mockBook} />
        </MemoryRouter>
      </Provider>
    );
    const readToggler = await screen.findByTestId("read__btn");
    fireEvent.click(readToggler);
  });
});
