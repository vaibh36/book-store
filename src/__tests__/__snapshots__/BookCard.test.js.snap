// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard Component snapshot test case 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-iqsm6x-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1lz8b89-MuiTypography-root"
          >
            Test Book
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
          >
            by 
            Test Author
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
          >
            Unread
          </p>
          <div
            class="MuiStack-root css-vuohop-MuiStack-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedDefault MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorDefault MuiButton-root MuiButton-contained MuiButton-containedDefault MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorDefault css-ladjc8-MuiButtonBase-root-MuiButton-root"
              data-testid="read__btn"
              tabindex="0"
              type="button"
            >
              Mark as Read
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-adro6z-MuiButtonBase-root-MuiIconButton-root"
              data-testid="info__icon"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="InfoIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
                />
              </svg>
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-adro6z-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-iqsm6x-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1lz8b89-MuiTypography-root"
        >
          Test Book
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
        >
          by 
          Test Author
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
        >
          Unread
        </p>
        <div
          class="MuiStack-root css-vuohop-MuiStack-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedDefault MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorDefault MuiButton-root MuiButton-contained MuiButton-containedDefault MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorDefault css-ladjc8-MuiButtonBase-root-MuiButton-root"
            data-testid="read__btn"
            tabindex="0"
            type="button"
          >
            Mark as Read
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-adro6z-MuiButtonBase-root-MuiIconButton-root"
            data-testid="info__icon"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="InfoIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
              />
            </svg>
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-adro6z-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
